aptos-move/framework/aptos-framework/sources/overloadable_fungible_asset.move

```diff
@@ -0,0 +1,169 @@
+    const EALREADY_REGISTERED: u64 = 5;
+    /// Fungibility is only available for non-deletable objects.
+    const EOBJECT_IS_DELETABLE: u64 = 18;
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct OverloadFunctionStore has key {
+		withdraw_function: FunctionInfo,
+		deposit_function: FunctionInfo,
+        transfer_ref: TransferRef,
+    }
+
+    /// Create a fungible asset store whose transfer rule would be overloaded by the provided function.
+    public fun register_overload_functions(
+        constructor_ref: &ConstructorRef,
+        withdraw_function: FunctionInfo,
+		deposit_function: FunctionInfo,
+    ) {
+        let dispatcher_withdraw_function_info = function_info::new_function_info(
+	        @aptos_framework,
+            string::utf8(b"overloadable_fungible_asset"),
+            string::utf8(b"dispatchable_withdraw"),
+        );
+        // Verify that caller type matches callee type so wrongly typed function cannot be registered.
+        assert!(function_info::check_dispatch_type_compatibility(
+            &dispatcher_withdraw_function_info,
+            &withdraw_function
+        ), error::invalid_argument(EWITHDRAW_FUNCTION_SIGNATURE_MISMATCH));
+
+        let dispatcher_deposit_function_info = function_info::new_function_info(
+	        @aptos_framework,
+            string::utf8(b"overloadable_fungible_asset"),
+            string::utf8(b"dispatchable_deposit"),
+        );
+        // Verify that caller type matches callee type so wrongly typed function cannot be registered.
+        assert!(function_info::check_dispatch_type_compatibility(
+            &dispatcher_deposit_function_info,
+            &deposit_function
+        ), error::invalid_argument(EDEPOSIT_FUNCTION_SIGNATURE_MISMATCH));
+
+        assert!(!object::can_generate_delete_ref(constructor_ref), error::invalid_argument(EOBJECT_IS_DELETABLE));
+        assert!(!exists<OverloadFunctionStore>(object::address_from_constructor_ref(constructor_ref)), error::already_exists(EALREADY_REGISTERED));
+
+        // Freeze the FungibleStore to force usign the new overloaded api.
+        let extend_ref = object::generate_extend_ref(constructor_ref);
+        fungible_asset::set_global_frozen_flag(&extend_ref, true);
+
+        let store_obj = &object::generate_signer(constructor_ref);
+
+        // Store the overload function hook.
+        move_to<OverloadFunctionStore>(store_obj, OverloadFunctionStore {
+            withdraw_function,
+		    deposit_function,
+            transfer_ref: fungible_asset::generate_transfer_ref(constructor_ref),
+        });
+    }
+
+    /// Withdraw `amount` of the fungible asset from `store` by the owner.
+    ///
+    /// The semantics of deposit will be governed by the function specified in OverloadFunctionStore.
+    public fun withdraw<T: key>(
+        owner: &signer,
+        store: Object<T>,
+        amount: u64,
+    ): FungibleAsset acquires OverloadFunctionStore {
```

Comment:
This should be a wrapper around withdraw, the happy path is that the FA isn't frozen and that this isn't used. We'll also need to see performance of the global freeze.